lib/Sema/CSSimplify.cpp

```diff
@@ -10418,33 +10420,28 @@ void ConstraintSystem::addContextualConversionConstraint(
                 convertTypeLocator, /*isFavored*/ true);
 }
 
-Type ConstraintSystem::addJoinConstraint(
-    ConstraintLocator *locator,
-    ArrayRef<std::pair<Type, ConstraintLocator *>> inputs) {
-  switch (inputs.size()) {
-  case 0:
-    return Type();
-
-  case 1:
-    return inputs.front().first;
+Optional<ExprKind>
+ConstraintSystem::getAtomicLiteralKind(TypeVariableType *typeVar) const {
+  const std::unordered_set<ExprKind> atomicLiteralKinds = {
+    ExprKind::IntegerLiteral,
+    ExprKind::FloatLiteral,
+    ExprKind::StringLiteral,
+    ExprKind::BooleanLiteral,
+    ExprKind::NilLiteral,
+  };
 
-  default:
-    // Produce the join below.
-    break;
-  }
+  if (!typeVar)
+    return None;
 
-  // Create a type variable to capture the result of the join.
-  Type resultTy = createTypeVariable(locator,
-                                     (TVO_PrefersSubtypeBinding |
-                                      TVO_CanBindToNoEscape));
+  auto *locator = typeVar->getImpl().getLocator();
+  if (!locator->directlyAt<LiteralExpr>())
+    return None;
 
-  // Introduce conversions from each input type to the type variable.
-  for (const auto &input : inputs) {
-    addConstraint(
-      ConstraintKind::Conversion, input.first, resultTy, input.second);
-  }
+  auto literalKind = getAsExpr(locator->getAnchor())->getKind();
```

Comment:
I think we could simplify all of this into `return castToExpr<LiteralExpr>(locator->getAnchor())->getKind();` and remove a set.

Reply:
Even better we could add that as a method on `TypeVariableType::Implementation` just like we have it for generic parameter and other things...

Reply:
I think we should just make it a switch instead of `unordered_set` because that would allocate heap memory on every call...

Reply:
The reason why I implemented this using a `std::unordered_set` is because I don't want to include _all_ literal expressions. For example, the set does not include `InterpolatedStringLiteralExpr` or `ObjectLiteralExpr`

Reply:
We could also make the `unordered_set` static so it's not allocated on every call. A switch would be fine too

Reply:
Yeah, let's just make it a switch over what we want here, I didn't realize that there is `NumericLiteral` which is abstract kind of all numeric literals...

Reply:
`NumberLiteral` is an abstract `Expr` class but the `ExprKind` is different for the different subclasses (e.g. `ExprKind::IntegerLiteral` and `ExprKind::FloatLiteral`)

Reply:
✅